File: user-service/src/main/java/com/vb/springboot/user/controller/UsersController.java
package com.vb.springboot.user.controller;

import com.vb.springboot.user.shared.UserDto;
import com.vb.springboot.user.model.request.UpdateUserDetailsRequestModel;
import com.vb.springboot.user.model.request.CreateUserRequestModel;
import com.vb.springboot.user.model.response.UserRestResponse;
import com.vb.springboot.user.service.UsersService;
import org.modelmapper.ModelMapper;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.core.env.Environment;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import javax.validation.Valid;

@RestController
@RequestMapping("/users")
public class UsersController {

    private final UsersService usersService;

    private final Environment environment;

    private final ModelMapper modelMapper;


    @Autowired
    public UsersController(final UsersService usersService,
                           final Environment environment,
                           final ModelMapper modelMapper) {
        this.usersService = usersService;
        this.environment = environment;
        this.modelMapper = modelMapper;
    }

    public String getUsers(@RequestParam(value = "page", defaultValue = "1") int page,
                           @RequestParam(value = "limit", defaultValue = "50") int limit,
                           @RequestParam(value = "sort", defaultValue = "desc", required = false) String sort) {
        return "get users was called with page = " + page + " and limit = " + limit + " and sort = " + sort;
    }

    @GetMapping(path = {"/{userId}"},
            produces = {MediaType.APPLICATION_JSON_VALUE, MediaType.APPLICATION_XML_VALUE})
    public ResponseEntity<UserRestResponse> getUser(@PathVariable String userId) {
        UserRestResponse userRestResponse = usersService.getUser(userId);

        if (userRestResponse == null) {
            return new ResponseEntity<>(HttpStatus.NO_CONTENT);
        }

        return new ResponseEntity<>(userRestResponse, HttpStatus.OK);
    }

    @PostMapping(consumes = {MediaType.APPLICATION_JSON_VALUE, MediaType.APPLICATION_XML_VALUE},
                 produces = {MediaType.APPLICATION_JSON_VALUE, MediaType.APPLICATION_XML_VALUE})
    public ResponseEntity<UserRestResponse> createUser(@Valid @RequestBody CreateUserRequestModel user) {
        UserDto userDto = modelMapper.map(user, UserDto.class);

        UserDto savedUser = usersService.createUser(userDto);

        return new ResponseEntity<>(modelMapper.map(savedUser, UserRestResponse.class), HttpStatus.OK);
    }

    @PutMapping(path = {"/{userId}"},
            consumes = {MediaType.APPLICATION_JSON_VALUE, MediaType.APPLICATION_XML_VALUE},
            produces = {MediaType.APPLICATION_JSON_VALUE, MediaType.APPLICATION_XML_VALUE})
    public ResponseEntity<UserRestResponse> updateUser(@PathVariable String userId,
                                                       @RequestBody UpdateUserDetailsRequestModel userToUpdate) {
        UserRestResponse storedUser = usersService.updateUser(userId, userToUpdate);
        return new ResponseEntity<>(storedUser, HttpStatus.OK);
    }

    @DeleteMapping(path = {"/{userId}"})
    public ResponseEntity<Void> deleteUser(@PathVariable String userId) {
        usersService.deleteUser(userId);

        return ResponseEntity.noContent().build();
    }

    @GetMapping("/status/check")
    public String status() {
        return "Working ))) Hello from USERS-WS PORT: " + environment.getProperty("local.server.port");
    }

}

File: user-service/src/main/kotlin/com/vb/springboot/user/controller/UsersController.kt
package com.vb.springboot.user.controller

import com.vb.springboot.user.model.request.CreateUserRequest
import com.vb.springboot.user.model.request.UpdateUserDetailsRequestModel
import com.vb.springboot.user.model.response.CreateUserResponse
import com.vb.springboot.user.service.UsersService
import com.vb.springboot.user.shared.UserDto
import org.modelmapper.ModelMapper
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.core.env.Environment
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import javax.validation.Valid

@RestController
@RequestMapping("/users")
class UsersController @Autowired constructor(private val usersService: UsersService,
                                             private val environment: Environment,
                                             private val modelMapper: ModelMapper) {

    fun getUsers(@RequestParam(value = "page", defaultValue = "1") page: Int,
                 @RequestParam(value = "limit", defaultValue = "50") limit: Int,
                 @RequestParam(value = "sort", defaultValue = "desc", required = false) sort: String): String {
        return "get users was called with page = $page and limit = $limit and sort = $sort"
    }

    @GetMapping(path = ["/{userId}"], produces = [MediaType.APPLICATION_JSON_VALUE, MediaType.APPLICATION_XML_VALUE])
    fun getUser(@PathVariable userId: String): ResponseEntity<CreateUserResponse> {
        val user = usersService.getUser(userId)
                ?: return ResponseEntity(HttpStatus.NO_CONTENT)
        return ResponseEntity(modelMapper.map(user, CreateUserResponse::class.java), HttpStatus.OK)
    }

    @PostMapping(consumes = [MediaType.APPLICATION_JSON_VALUE, MediaType.APPLICATION_XML_VALUE],
                 produces = [MediaType.APPLICATION_JSON_VALUE, MediaType.APPLICATION_XML_VALUE])
    fun createUser(@Valid @RequestBody user: CreateUserRequest): ResponseEntity<CreateUserResponse> {
        val userDto = modelMapper.map(user, UserDto::class.java)
        val savedUser = usersService.createUser(userDto)
        return ResponseEntity(modelMapper.map(savedUser, CreateUserResponse::class.java), HttpStatus.OK)
    }

    @PutMapping(path = ["/{userId}"],
            consumes = [MediaType.APPLICATION_JSON_VALUE, MediaType.APPLICATION_XML_VALUE],
            produces = [MediaType.APPLICATION_JSON_VALUE, MediaType.APPLICATION_XML_VALUE])
    fun updateUser(@PathVariable userId: String,
                   @RequestBody userToUpdate: UpdateUserDetailsRequestModel): ResponseEntity<CreateUserResponse> {
        val userDto = modelMapper.map(userToUpdate, UserDto::class.java)

        val updatedUser = usersService.updateUser(userId,  userDto)
                ?: return ResponseEntity(HttpStatus.NO_CONTENT)
        return ResponseEntity(modelMapper.map(updatedUser, CreateUserResponse::class.java), HttpStatus.CREATED)
    }

    @DeleteMapping(path = ["/{userId}"])
    fun deleteUser(@PathVariable userId: String): ResponseEntity<Void> {
        usersService.deleteUser(userId)
        return ResponseEntity.noContent().build()
    }

    @GetMapping("/status/check")
    fun status(): String {
        return "Working ))) Hello from USERS-WS PORT: " + environment.getProperty("local.server.port")
    }
}